File: 383-ransom-note/ransom-note.java
class Solution {
    public boolean canConstruct(String ransomNote, String magazine) {
        HashMap<Character, Integer> dictionary = new HashMap<>();

        for(Character c: magazine.toCharArray()) {
            dictionary.put(c, dictionary.getOrDefault(c, 0) + 1);
        }
        // System.out.println(dictionary);
        for(Character c: ransomNote.toCharArray()) {
            if(dictionary.containsKey(c) && dictionary.get(c) > 0){
                dictionary.put(c, dictionary.get(c)-1);
            }
            else {
                return false;
            }
        }
        return true;
    }
}

File: 383-ransom-note/ransom-note.py
class Solution:
    def canConstruct(self, ransomNote: str, magazine: str) -> bool:

        #Brute Force
        '''
        iterate over ransom Note
        check if each letter exists in magzine
        the second step will be O(n) in brute force but if we use 
        hashmap, the lookup will be O(1)
        '''

        lookup = defaultdict(int)
        for char in magazine:
            lookup[char]+=1

        # print(lookup)
        
        for char in ransomNote:
            if char in lookup and lookup[char] != 0:
                lookup[char] -=1
            else:
                return False
        
        
        return True

   

        